### DISCUSSION TITLE
Avalonia for Text-based user interface hosted by console

### DISCUSSION THREAD

**Discussion by jinek** [AvaloniaUI/Avalonia#6274] [General]
Hey Avalonians!
I'm going to implement avalonia drawing and windowing abstraction for [TUI](https://en.wikipedia.org/wiki/Text-based_user_interface).
.NET already has a [framework](https://github.com/elw00d/consoleframework), but it has its own DependencyObject/Layout/Xml parser which all together don't have so big support as avalonia.

I've already tried to create small demo with two drawing primitives: glyph and rectangle:
![image](https://user-images.githubusercontent.com/10516222/126203330-7cdc3b3f-45f4-4963-9f6a-e67f194553ad.png)
I need to learn a lot about Avalonia rendering system, also I need to create special theme (FontSize=1, no transparancy, smaller chrome sizes, less borders etc) and all these seems not small work for me.

May be anybody interested in the topic? I would appreciate any help here. Especially would be perfect to have any consulting from people who understand drawing abstraction to help planning the implementation.

Additionally, may be anybody have deep knowledges about terminals, like Windows Terminal, various linux terminals and others under platforms which can run .net 5 apps?

Few inspirational pics on the topic
![image](https://user-images.githubusercontent.com/10516222/126206761-b6084aee-4b4a-49f9-8d70-7e5f963481a8.png)
![image](https://user-images.githubusercontent.com/10516222/126206870-bba1f74f-4382-479e-844f-f9ba10b928c8.png)
Vim ofcoz
![image](https://user-images.githubusercontent.com/10516222/126207477-b8b868a9-c88d-41ed-8723-7086f9dc23fe.png)
Github cat translated to ANSI (example of drawing images)
![image](https://user-images.githubusercontent.com/10516222/126209084-0a0a4586-386d-4a82-903b-fc47cd65d349.png)

**Comment by FoggyFinder** [comment]
But...why? Aren't there are enough libraries for TUI? Honestly I can't see how Avalonia can help with it

**Reply by jinek** [reply]
Avalonia has DependencyObject/Property, Binding, Layout, XAML parser, ItemsList virtualization, 3d party controls, Animation etc, which other libraries don't have (or don't have so well supported).

**Comment by kekekeks** [comment]
You'll probably need your own IRenderer implementation that would ignore `Render` method completely and interpret controls in a different way. You'll also need to replace IPlatformRenderInterface with something which would provide fixed size font metrics.

**Reply by jinek** [reply]
Thank you!
1) Why shouldn't I draw within `Render` method?
2) Can't I just re-use existing `ImmediateRenderer` (or differed one) ? It already has logic to run the visual tree and call `IVisual.Render`, same time it gives possibiliy to use custom implementation when it calls to ```using (var context = new DrawingContext(target.CreateDrawingContext(renderer)))```. I could just return my custom drawing context instead.
3) When should I run the drawing? I receieve `Invalidate` call, can I just accumulate all this calls maden within one Dispatcher cycle and run drawing on next cycle?

**Reply by kekekeks** [reply]
The problem is, our rendering infrastructure is designed to work with pixels, while you need pseudographic unicode symbols. So I'd advise to have a separate rendering infrastructure designed with fixed symbol grid in mind instead.

**Comment by jinek** [comment]
I have developed a proof of concept. It has some drawing and control styles and I'm looking for support.
Please, consider if you could be interested in the project:
https://github.com/jinek/Consolonia

![](https://raw.githubusercontent.com/jinek/Consolonia/main/showcase.gif)

**Reply by robloo** [reply]
Wow, I won't use this personally but the fact it was possible is quite amazing. I've never seen some of these ideas. Great job on implementing this!
